File: lua/youtube_bz/brainz.lua
#!/usr/bin/env lua
-- brainz.lua
module(..., package.seeall)

-- Require
local http_request = require "http.request"
local json = require "json"

-- Function

---[[ Get Brainz result query
function getBrainzResult(mbid)
        local url = string.format("https://musicbrainz.org/ws/2/release/%s?inc=artists+recordings&fmt=json", mbid)
        local headers, stream = assert(http_request.new_from_uri(url):go())
        local body = assert(stream:get_body_as_string())
        if headers:get ":status" ~= "200" then
            error(body)
            return
        end
        return json.decode(body)
end
--]]

File: golang/youtube_bz/brainz.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

type Release struct {
	Title        string         `json:"title"`
	ArtistCredit []ArtistCredit `json:"artist-credit"`
	Media        []Media        `json:"media"`
}

type ArtistCredit struct {
	Name string `json:"name"`
}

type Media struct {
	Tracks []Track `json:"tracks"`
}

type Track struct {
	Title  string `json:"title"`
	Length int    `json:"length"`
}

func brainz_search(mbid string) Release {
	var url string = fmt.Sprintf("https://musicbrainz.org/ws/2/release/%s?inc=artists+recordings&fmt=json", mbid)
	resp, err := http.Get(url)
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}

	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}

	jsonBlob := string(body)
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}

	var release Release
	json.Unmarshal([]byte(jsonBlob), &release)

	return release
}
